### PR TITLE
Drop support for Python 3.8

### PR DESCRIPTION
**Pull request by arcondello**
Hopefully the typing are all drop-in replacements. Though huge find/replaces are fraught.

### CODE REVIEW

File: dimod/constrained/constrained.py [hunk 1]
```diff
@@ -1472,7 +1471,7 @@ def count(qm: Union[QuadraticModel, BinaryQuadraticModel]) -> int:
 
         return n
 
-    def relabel_constraints(self, mapping: Mapping[Hashable, Hashable]):
+    def relabel_constraints(self, mapping: Mapping[collections.abc.Hashable, collections.abc.Hashable]):
```
Comment (thisac) [inline dimod/constrained/constrained.py:1474]:
```suggestion
    def relabel_constraints(self, mapping: collections.abc.Mapping[collections.abc.Hashable, collections.abc.Hashable]):
```

Comment (thisac) [reply dimod/constrained/constrained.py:1474]: There are a few more in the `CQMToBQMInverter` class.

File: dimod/constrained/cyexpression.pyx [hunk 2]
```diff
@@ -12,8 +12,8 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 
+import collections.abc
```
Comment (thisac) [inline dimod/constrained/cyexpression.pyx:15]:
```suggestion
```
Not used it seems.

File: dimod/quadratic/quadratic_model.py [hunk 3]
```diff
@@ -1106,12 +1105,12 @@ def reduce_quadratic(self, function: Callable,
...
         return self.data.reduce_quadratic
 
-    def relabel_variables(self, mapping: Mapping[Variable, Variable],
+    def relabel_variables(self, mapping: collections.abc.Mapping[Variable, Variable],
                           inplace: bool = True) -> 'QuadraticModel':
         """Relabel the variables according to the given mapping.
 
         Args:
-            mapping: Mapping of current variable labels to new ones. If an
+            mapping: collections.abc.Mapping of current variable labels to new ones. If an
```
Comment (thisac) [inline dimod/quadratic/quadratic_model.py:1113]:
```suggestion
            mapping: Mapping of current variable labels to new ones. If an
```

File: dimod/constrained/constrained.py (hunk 1 above)
Comment (arcondello) [reply dimod/constrained/constrained.py]: Think I got them all in 49f92ba. Gosh I hate Python's `typing`, especially how unstable it's been over the last few versions.

**Review by thisac — APPROVED** [review]
LGTM

**Review by randomir — APPROVED** [review]
gotta love typing in python :exploding_head:

**Pull request merged**